server/filestore.go

```diff
@@ -5295,6 +5313,45 @@ func (fs *fileStore) cancelAgeChk() {
 	}
 }
 
+// Lock must be held so that nothing else can interleave and write a
+// new message on this subject before we get the chance to write the
+// system tombstone. If the tombstone is written successfully then
+// this function returns a callback func to call scb and tcb after the
+// lock has been released.
+func (fs *fileStore) systemTombstoneIfNeeded(sm *StoreMsg, reason string) func() {
+	// If the deleted message was itself a tombstone then don't
+	// write out more tombstones or we'll churn endlessly.
+	if len(getHeader(JSAppliedLimit, sm.hdr)) != 0 {
+		return nil
+	}
+	if !fs.cfg.AllowMsgTTL || fs.cfg.LimitsTTL <= 0 {
```

Comment:
Be better to do this check in upper sections and not do the unlock / lock in removeMsg.